### PR TITLE
More histogram support, redo stats table

### PR DESCRIPTION
**Pull request by max-hoffman**
Reshape statistics interfaces to better support custom Dolt implementation.

- Add `ANALYZE TABLE <table> [UPDATE/DROP] HISTOGRAM ON <column,...> USING <json blob>` support and a few tests
- Replace use of `update information schema` cardinality updates with histogram updates
- Default catalog has an in-memory histogram
- New memory histogram impl
- Delete old statistics table/histogram code

The only prod difference is that the new update path can overwrite the default table count, while it was a testing-only thing before.

companion: https://github.com/dolthub/vitess/pull/279

Dolt bump seems OK

### CODE REVIEW

File: memory/stats.go [hunk 1]
```diff
@@ -0,0 +1,256 @@
...
+			Columns:   cols,
+			Types:     types,
+			Version:   0,
+			CreatedAt: time.Now(),
+		}
+	}
+	return nil
+}
+
+func (s *StatsProv) reservoirSample(ctx *sql.Context, table sql.Table) ([]sql.Row, error) {
```
Comment (zachmu) [inline memory/stats.go:145]: comment on the algorithm used here would be nice

File: sql/planbuilder/analyze.go [hunk 2]
```diff
@@ -15,36 +15,135 @@
...
 	outScope = inScope.push()
-	names := make([]sql.DbTable, len(n.Tables))
 	defaultDb := b.ctx.GetCurrentDatabase()
-	for i, table := range n.Tables {
-		dbName := table.Qualifier.String()
-		if dbName == "" {
-			if defaultDb == "" {
-				err := sql.ErrNoDatabaseSelected.New()
+
+	if n.Action == "" {
```
Comment (zachmu) [inline sql/planbuilder/analyze.go:34]: Recommend pulling out methods for these long case bodies

File: sql/statistics.go [hunk 3]
```diff
@@ -29,60 +28,12 @@ type StatisticsTable interface {
...
-	// Histograms returns a map from all column names to their associated histograms.
-	Histograms HistogramMap
-}
-
-func (ts *TableStatistics) Histogram(colName string) (*Histogram, error) {
-	if res, ok := ts.Histograms[colName]; ok {
-		return res, nil
-	}
-	return &Histogram{}, fmt.Errorf("column %s not found", colName)
+type StatsProvider interface {
```
Comment (zachmu) [inline sql/statistics.go:31]: Need comments on these methods

File: sql/stats/histogram.go [hunk 4]
```diff
@@ -0,0 +1,3 @@
+package stats
```
Comment (zachmu) [inline sql/stats/histogram.go:1]: Delete this for now?

File: enginetest/join_planning_tests.go [hunk 5]
```diff
@@ -55,7 +55,10 @@ var JoinPlanningTests = []struct {
 			"insert into rs values (0,0), (1,0), (2,0), (4,4), (5,4);",
 			"insert into uv values (0,1), (1,1), (2,2), (3,2);",
 			"insert into ab values (0,2), (1,2), (2,2), (3,1);",
-			"update information_schema.statistics set cardinality = 1000 where table_name in ('ab', 'rs', 'xy', 'uv');",
+			`analyze table xy update histogram on x using '{"row_count":1000}'`,
```
Comment (zachmu) [inline enginetest/join_planning_tests.go:58]: Need to test the error case: giving a non-json data value, or one that doesn't have the right fields

**Review by zachmu — APPROVED** [review]
Looks pretty reasonable

See comments on vitess PR about missing keyword

File: memory/stats.go [hunk 6]
```diff
@@ -0,0 +1,256 @@
...
+type statsKey string
+
+type StatsProv struct {
+	colStats   map[statsKey]*stats.Stats
+	dbProvider sql.DatabaseProvider
+}
+
+var _ sql.StatsProvider = (*StatsProv)(nil)
+
+func (s *StatsProv) RefreshTableStats(ctx *sql.Context, table sql.Table, db string) error {
```
[the remark this answers is not in the document]
Comment (max-hoffman) [reply memory/stats.go:32]: i suppose I could add that to the sql.Table interface

[the remark this answers is not in the document]
Comment (max-hoffman) [reply memory/stats.go:32]: nvm that's hairy, lots of non-table things implementing this

**Pull request merged**